packages/vitest/src/integrations/coverage/istanbul.ts

```diff
@@ -0,0 +1,143 @@
+import { existsSync, promises as fs } from 'fs'
+import { createRequire } from 'module'
+import { resolve } from 'pathe'
+import type { ExistingRawSourceMap, TransformPluginContext } from 'rollup'
+
+import { configDefaults, defaultExclude, defaultInclude } from '../../defaults'
+import type { Vitest } from '../../node'
+import type { IstanbulOptions, ResolvedCoverageOptions } from '../../types'
+import type { BaseCoverageReporter } from './base'
+
+const require = createRequire(import.meta.url)
+const coverageVariable = '__VITEST_COVERAGE__'
+
+interface Instrumenter {
+  /* Instrument the supplied code and track coverage against the supplied filename. It throws if invalid code is passed to it. ES5 and ES6 syntax is supported. To instrument ES6 modules, make sure that you set the esModules property to true when creating the instrumenter. */
+  instrumentSync(
+    /* The code to instrument */
+    code: string,
+    /* The filename against which to track coverage. */
+    filename: string,
+    /* The source map that maps the not instrumented code back to it's original form. Is assigned to the coverage object and therefore, is available in the json output and can be used to remap the coverage to the untranspiled source.): string; */
+    inputSourceMap: object
+  ): string
+
+  /* Returns the file coverage object for the last file instrumented. */
+  lastSourceMap(): ExistingRawSourceMap
+}
+
+interface TestExclude {
+  new(opts: {
+    cwd?: string | string[]
+    include?: string | string[]
+    exclude?: string | string[]
+    extension?: string | string[]
+    excludeNodeModules?: boolean
+  }): { shouldInstrument(filePath: string): boolean }
+}
+
+export class IstanbulReporter implements BaseCoverageReporter {
+  ctx!: Vitest
+  options!: ResolvedCoverageOptions & { provider: 'istanbul' }
+  instrumenter!: Instrumenter
+  testExclude!: InstanceType<TestExclude>
+  coverages: any[] = []
+
+  initialize(ctx: Vitest) {
+    this.ctx = ctx
+    this.options = resolveIstanbulOptions(ctx.config.coverage, ctx.config.root)
+
+    const { createInstrumenter } = require('istanbul-lib-instrument')
+    this.instrumenter = createInstrumenter(this.options)
+
+    const TestExclude = require('test-exclude')
+
+    this.testExclude = new TestExclude({
+      cwd: ctx.config.root,
+      exclude: [...defaultExclude, ...defaultInclude],
+      extension: ['.js', '.cjs', '.mjs', '.ts', '.tsx', '.jsx', '.vue', '.svelte'],
+      excludeNodeModules: true,
+    })
+  }
+
+  resolveOptions(): ResolvedCoverageOptions {
+    return this.options
+  }
+
+  onFileTransform(sourceCode: string, id: string, pluginCtx: TransformPluginContext) {
+    if (!this.testExclude.shouldInstrument(id))
+      return
+
+    // TODO: Sourcemaps are not accurate.
```

Comment:
Reminder for later: The key is to use `istanbul-lib-source-maps` here. Sourcemaps even for TS seem to be correct with that one. 🎉